GeneratorInterface/GenFilters/plugins/MCMultiParticleMassFilter.cc

```diff
@@ -0,0 +1,136 @@
+#include "GeneratorInterface/GenFilters/plugins/MCFilterZboostHelper.h"
+#include "DataFormats/Common/interface/Handle.h"
+#include "FWCore/Framework/interface/global/EDFilter.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/EDGetToken.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "SimDataFormats/GeneratorProducts/interface/HepMCProduct.h"
+
+#include <vector>
+#include <iostream>
+
+class MCMultiParticleMassFilter : public edm::global::EDFilter<> {
+public:
+  explicit MCMultiParticleMassFilter(const edm::ParameterSet&);
+  ~MCMultiParticleMassFilter() override;
+
+private:
+  bool filter(edm::StreamID, edm::Event& iEvent, const edm::EventSetup&) const override;
+  bool recurseLoop(std::vector<HepMC::GenParticle*>& particlesThatPassCuts, std::vector<int>& indices, int depth) const;
+
+  /* Member data */
+  const edm::EDGetTokenT<edm::HepMCProduct> token_;
+  const std::vector<int> particleID;
+  double ptMin;
+  double etaMax;
+  int status;
+  const double minTotalMassSq;
+  const double maxTotalMassSq;
+  int nParticles;
+
+  int particleSumTo;
+  int particleProdTo;
+};
+
+using namespace edm;
+using namespace std;
+
+MCMultiParticleMassFilter::MCMultiParticleMassFilter(const edm::ParameterSet& iConfig)
+    : token_(consumes<edm::HepMCProduct>(
+          iConfig.getUntrackedParameter<edm::InputTag>("src", edm::InputTag("generator", "unsmeared")))),
+      particleID(iConfig.getParameter<std::vector<int>>("ParticleID")),
+      ptMin(iConfig.getParameter<double>("PtMin")),
+      etaMax(iConfig.getParameter<double>("EtaMax")),
+      status(iConfig.getParameter<int>("Status")),
+      minTotalMassSq(iConfig.getParameter<double>("minTotalMass") * iConfig.getParameter<double>("minTotalMass")),
+      maxTotalMassSq(iConfig.getParameter<double>("maxTotalMass") * iConfig.getParameter<double>("maxTotalMass")) {
```

Comment:
I have done this. In order to facilitate using vectors, I have now included a mapping from the particle ID provided to its respective pt, eta, and status requirements.

For example, if one was to input:
```python
cms.EDFilter("MCMultiParticleMassFilter",
    EtaMax = cms.vdouble(2.7, 2.7, 3.0, 3.0),
    ParticleID = cms.vint32(-11, 11, -13, 13),
    PtMin = cms.vdouble(3.0, 3.0, 1.8, 1.8),
    Status = cms.vint32(1),
    minTotalMass = cms.double(100.0),
    maxTotalMass = cms.double(1e15),
)
```

It would say that the two electrons need a maximum eta of 2.7, and the two muons a maximum eta of 3. It would also say that the minimum pt of the electrons is 3, while for the muons it is 1.8. The status is set to a default of 1 throughout.


The one downside to the way that I have framed the code is that you _cannot_ have particles of the same id with different requirements. If I had something that looked like

```python
cms.EDFilter("MCMultiParticleMassFilter",
    EtaMax = cms.vdouble(2.7, 3.0),
    ParticleID = cms.vint32(11, 11),
    PtMin = cms.vdouble(3.0, 1.8),
    Status = cms.vint32(1),
    minTotalMass = cms.double(100.0),
    maxTotalMass = cms.double(1e15),
)
```

It would just set the requirements to be 2 electrons that add up to > 100 GeV in mass,